### PR TITLE
Update to TypeScript 4.1.3

### PR DESCRIPTION
**Pull request by jasongrout**
## References

What's new: https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/

## Code changes

The update required changing two things in the code, plus fixing one actual bug that it caught.

Putting as 3.0, under the assumption that since TS upgrades are breaking, we won't be able to change the TS version until 4.0, so we should really try to have the latest when we ship 3.0. TS 4.1 has been out almost a month now, and brings some nice things.

### CODE REVIEW

**Comment by jasongrout** [comment]
It looks like typedoc 0.20 beta now may support the library mode that was keeping us on 0.17.0-3: see https://github.com/TypeStrong/typedoc/pull/1184#issuecomment-719972623

File: typedoc.js [hunk 1]
```diff
@@ -22,9 +22,9 @@ module.exports = {
 
     '**/packages/metapackage/**'
   ],
-  excludeNotExported: true,
-  ignoreCompilerErrors: false,
-  mode: 'library',
+  // excludeNotExported: true,
+  // ignoreCompilerErrors: false,
+  // mode: 'library',
```
Comment (telamonian) [inline typedoc.js:27]: you can delete this line, at least. From what I read I don't think the typedoc guy wants to keep the name "library mode" around in the new version

**Comment by telamonian** [comment]
thanks for doing the typedoc upgrade. I've been following along with the progress, but I wasn't sure to what extent v0.20 would be a drop-in replacement for library mode from v0.17, so I hadn't gotten around to doing this update myself. Hopefully with all of the good work that @Gerrit0 has been doing there will still be more improvement to come for our API docs

File: typedoc.js (hunk 1 above)
Comment (jasongrout) [reply typedoc.js:27]: > you can delete this line, at least. From what I read I don't think the typedoc guy wants to keep the name "library mode" around in the new version

From the issue linked above, I think all three of these options are not in 0.20 anymore. Thanks for the confirmation.

**Comment by jasongrout** [comment]
> thanks for doing the typedoc upgrade

Thanks for jumping in here. The upgrade is certainly not done yet. It still is giving lots of errors. I spent a lot of time trying to wrangle the configuration to ignore testutils and other files that were having errors, but was not successful. I was planning to reach out to you tomorrow to see if you could help, since it seems you have more experience with typedoc these days. Do you want to take over getting typedoc to work from here? Or maybe we can merge this and get typedoc working better in another PR?

If we can't get typedoc working globally for all modules at once, it does appear that we can perhaps generate api docs for each package individually. I'm at the point where I'm giving up for now on getting it to work as a single all-encompassing build, but would rather at this point try to patch together individual package docs. If you can get the single all-encompassing build to work, that would be wonderful.

**Comment by jasongrout** [comment]
> The upgrade is certainly not done yet.

To be clear, I think the typescript upgrade is done and works great. The typedoc upgrade is not working yet.

**Comment by telamonian** [comment]
ooo, I had taken all of your typedoc-related commits to mean that we were already living in the happy space-future, but I guess not yet. Does the current v0.17-based typedoc build still work, or did that fall over when you upgraded typescript?

For now, let's move all of your typedoc commits to a new PR. I can take over the typedoc upgrade, but I definitely can't promise to get it done quickly. From what I've read of TypeStrong/typedoc#1364 the current v0.20 beta *should* be able to produce a unified api doc for our whole monorepo, but since stuff is still in flight on the typedoc side I can't even really promise that

**Comment by jasongrout** [comment]
> For now, let's move all of your typedoc commits to a new PR.

Sounds great. I've reverted this PR to just upgrading typescript, and opened #9483 for upgrading typedoc.

**Comment by jasongrout** [comment]
Note: @blink1073 requested that we not merge this until the api docs in #9483 is ready to go (so perhaps we should not have split it off to a different PR after all).

**Comment by jasongrout** [comment]
This has an extra lint commit beyond #9483, so merging this to pick up the lint change.

**Pull request merged**